features/cellular/framework/AT/ATHandler.cpp

```diff
@@ -337,48 +335,59 @@ void ATHandler::rewind_buffer()
     }
 }
 
-// we are always expecting to receive something so there is wait timeout
-void ATHandler::fill_buffer()
+int ATHandler::poll_timeout(bool wait_for_timeout)
+{
+    int timeout;
+    if (wait_for_timeout) {
+        uint64_t now = rtos::Kernel::get_ms_count();
+        if (now >= _start_time + _at_timeout) {
+            timeout = 0;
+        } else if ( _start_time + _at_timeout - now > INT_MAX) {
+            timeout = INT_MAX;
+        } else {
+            timeout = _start_time + _at_timeout - now;
+        }
+    } else {
+        timeout = 0;
+    }
+    return timeout;
+}
+
+bool ATHandler::fill_buffer(bool wait_for_timeout)
 {
     tr_debug("%s", __func__);
     // Reset buffer when full
     if (sizeof(_recv_buff) == _recv_len) {
         reset_buffer();
     }
 
-    Timer timer;
-    timer.start();
-    do {
-       ssize_t len = _fileHandle->read(_recv_buff + _recv_len, sizeof(_recv_buff) - _recv_len);
+    pollfh fhs;
+    fhs.fh = _fileHandle;
+    fhs.events = POLLIN;
+    int timeout = 0;
+    int count = poll(&fhs, 1, poll_timeout(wait_for_timeout));
+    if (count > 0 && (fhs.revents & POLLIN)) {
+        ssize_t len = _fileHandle->read(_recv_buff + _recv_len, sizeof(_recv_buff) - _recv_len);
         if (len > 0) {
             _recv_len += len;
-           at_debug("\n----------readable----------: %d\n", _recv_len);
-           for (size_t i = _recv_pos; i < _recv_len; i++) {
-               at_debug("%c", _recv_buff[i]);
-           }
-           at_debug("\n----------readable----------\n");
-           return;
-       } else if (len != -EAGAIN && len != 0) {
-           tr_warn("%s error: %d while reading", __func__, len);
-           break;
-       }
-#ifdef MBED_CONF_RTOS_PRESENT
-        rtos::Thread::yield();
-#endif
-    } while ((uint32_t)timer.read_ms() < _at_timeout);
+            return true;
+        }
+    }
 
-    set_error(NSAPI_ERROR_DEVICE_ERROR);
-    tr_debug("AT TIMEOUT, scope: %d timeout: %lu", _current_scope, _at_timeout);
+    if (wait_for_timeout) {
```

Comment:
It possibly feels like this `set_error` should be one layer up, where you're printing the `tr_warn("AT TIMEOUT")`. The error view maybe depends what you were waiting for.